doc/man/doveadm-replicator.1.in
.\" Copyright (c) 2014-2018 Dovecot authors, see the included COPYING file
.TH DOVEADM\-REPLICATOR 1 "2014-10-05" "Dovecot v2.3" "Dovecot"
.SH NAME
doveadm\-replicator \- Manage users\(aq mail replication
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.BR doveadm " [" \-Dv ]
[\fB\-f\fP \fIformatter\fP]
.BI replicator \ command
.RI [ OPTIONS ]\ [ ARGUMENTS ]
.\"------------------------------------------------------------------------
.SH DESCRIPTION
The
doveadm replicator
.I COMMANDS
can be used to manage the automated replication of users\(aq mail.
.\"------------------------------------------------------------------------
@INCLUDE:global-options-formatter@
.\" --- command specific options --- "/.
.PP
This command uses by default the output formatter
.B flow
(without the
.IR key =
prefix).
.PP
Command specific
.IR options :
.TP
.BI \-a \ replicator_socket_path
This option is used to specify an absolute path to an alternative UNIX
domain socket.
.sp
By default
.BR doveadm (1)
will use the socket
.IR @rundir@/replicator\-doveadm .
The socket may be located in another directory, when the default
.I base_dir
setting was overridden in
.IR @pkgsysconfdir@/dovecot.conf .
.\"------------------------------------------------------------------------
.SH ARGUMENTS
.TP
.I user_mask
Specifies for which users the command should be executed.
In most cases you would like to use patterns which contains wildcards,
e.g. \(rs*@example.net for all users of the example.net domain, or simply
\(rs* for all users.
.sp
When the SQL userdb module is used make sure that the
.I iterate_query
setting in
.I @pkgsysconfdir@/dovecot\-sql.conf.ext
matches your database layout.
When using the LDAP userdb module, make sure that the
.IR iterate_attrs " and " iterate_filter
settings in
.I @pkgsysconfdir@/dovecot\-ldap.conf.ext
match your LDAP schema.
Otherwise
.BR doveadm (1)
will be unable to iterate over all users.
.\"------------------------------------------------------------------------
.SH COMMANDS
.SS replicator add
.B doveadm replicator add
[\fB\-a\fP \fIreplicator_socket_path\fP]
.I user_mask
.PP
Add the specified user(s) to the replicator.
If the
.I user_mask
contains \(dq?\(dq or \(dq*\(dq wildcards, the list of usernames is looked
up from the
.IR userdb .
.\"-------------------------------------
.SS replicator dsync\-status
.B doveadm replicator dsync\-status
[\fB\-a\fP \fIreplicator_socket_path\fP]
.PP
Show the status for the currently running dsync processes.
.\"-------------------------------------
.SS replicator remove
.B doveadm replicator remove
[\fB\-a\fP \fIreplicator_socket_path\fP]
.I username
.PP
Remove the specified user from replicator.
.\"-------------------------------------
.SS replicator replicate
.B doveadm replicator replicate
[\fB\-a\fP \fIreplicator_socket_path\fP]
.RB [ \-f ]
[\fB\-p\fP \fIpriority\fP]
.I user_mask
.PP
Start replication for the specified users now.
If the \-f parameter is given, full replication is done for the user.
You can also specify the priority, which can be either
.B high " or " low.
If the user mask contains \(dq?\(dq or \(dq*\(dq wildcards, the list of
usernames is looked up from the users that currently exist in replicator
(not from the
.IR userdb ).
.\"-------------------------------------
.SS replicator status
.B doveadm replicator status
[\fB\-a\fP \fIreplicator_socket_path\fP]
.I user_mask
.PP
Show the replication status for users.
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH SEE ALSO
.BR doveadm (1),
.BR doveadm\-sync (1),
.\"-------------------------------------
.PP
Additional resources:
.IP "Replication with dsync"
https://doc.dovecot.org/configuration_manual/replication/


NEWS
v2.3.17.1 2021-12-07  Aki Tuomi <aki.tuomi@open-xchange.com>

	- dsync: Add back accidentically removed parameters.
	- lib-ssl-iostream: Fix assert-crash when OpenSSL returned syscall error
	  without errno.
	- master: Dovecot failed to start if ssl_ca was too large.

v2.3.17 2021-10-28  Aki Tuomi <aki.tuomi@open-xchange.com>

	* Dovecot now logs a warning if time seems to jump forward at least
	  100 milliseconds.
	* dict: Lines logged by the dict process now contain the dict name as
	  the prefix.
	* lib-index: mail_cache_fields, mail_always_cache_fields and
	  mail_never_cache_fields now verifies that the listed header names are
	  valid. Especially the UTF8 "–" character has sometimes been wrongly
	  used instead of the ASCII "-".
	+ *-login: Added login_proxy_rawlog_dir setting to capture
	  rawlogs between proxy and backend.
	+ dict: The server process now keeps the last 10 idle dict backends
	  cached for maximum of 30 seconds. Practically this acts as a
	  connection pool for dict-redis and dict-ldap. Note that this doesn't
	  affect dict-sql, because it already had its own internal cache.
	+ doveadm: New stats add/remove commands added to support changing the
	  metrics configuration on runtime.
	+ lazy_expunge: Added lazy_expunge_exclude settings to disable
	  lazy_expunge for specific folders. \Special-use flags can be used as
	  folder names.
	+ lib-lua: Added a new helper function dovecot.restrict_global_variables()
	  to disable or enable defining new global variables.
	- LAYOUT=index List index rebuild was missing.
	- LAYOUT=index: Duplicate GUIDs were not detected.
	- acl: When using acl_ignore_namespace Dovecot attempted to access or
	  create dovecot-acl-list even when the namespace should have been
	  ignored. For virtual namespaces this could have yielded errors about
	  "Read-only file system" or "Permission denied".
	- auth: Setting the "master" passdb field to empty value would
	  cause proxying to fail with an authentication error.
	  Now an empty "master" field is ignored.
	- doveadm-server: Duplicate error lines were sent for failed commands.
	  This didn't normally cause visible problems, except when using
	  wildcards in usernames or -A parameter to go through multiple users.
	- doveadm-server: Logs written by doveadm-server were often missing log
	  prefixes, especially mail_log_prefix for mail commands. Logs sent to
	  doveadm TCP client were also missing log prefixes.
	- doveadm: v2.3 regression: batch command always crashes.
	- doveadm: v2.3.11 regression: Commands failed if ssl_cert or
	  ssl_key files weren't readable by the user running doveadm, even
	  though doveadm didn't actually use these settings
	- imap-hibernate: Process may crash at deinit:
	  Panic: file ioloop.c: line 928 (io_loop_destroy): assertion failed:
	  (ioloop->cur_ctx == NULL).
	- imap: Using imap_fetch_failure=no-after can cause assert-crash
	  with some IMAP commands if reading the mail fails (e.g. wrong cached
	  mail size). Fixes:
	  Panic: file index-mail-headers.c: line 198 (index_mail_parse_header_init):
	  assertion failed: (!mail->data.header_parser_initialized)
	- imap: v2.3.10 regression: When using INDEXPVT to enable private
	  \Seen flags (for shared or public namespaces) the STORE command did
	  not send untagged replies for the \Seen flag changes.
	- imap: v2.3.15 regression: If PREVIEW/SNIPPET is not the final FETCH
	  option in the command, the IMAP FETCH response is broken.
	- imap: v2.3.15 regression: MOVE command leaks mailbox if it can't be
	  opened and crashes at deinit:
	  Panic: file mail-user.c: line 229 (mail_user_deinit): assertion failed:
	  ((*user)->refcount == 1).
	- imapc: Copying nonexistent mail via imapc could have crashed. Fixes:
	  Panic: file mail-storage.c: line 2385 (mailbox_transaction_commit_get_changes):
	  assertion failed: (ret < 0 || seq_range_count(&changes_r->saved_uids) == save_count ||
	  array_count(&changes_r->saved_uids) == 0).
	- indexer: v2.3.15 regression: Process crashes if indexer-client
	  disconnects while it's waiting for command reply. This happened for
	  example if IMAP SEARCH triggered long fts indexing and the IMAP
	  client disconnected while waiting for the reply.
	- indexer: v2.3.15 regression: Process may have crashed in some situations.
	- indexer: v2.3.15 regression: indexer-worker processes may not have
	  reached the process_limit in some situations, possibly even using just
	  one indexer-worker process even though there were many indexing
	  requests queued.
	- lib-compression: Reading lz4 compressed mdbox mails may crash. Fixes:
	  Panic: file istream.c: line 345 (i_stream_read_memarea):
	  assertion failed: (!stream->blocking).
	- lib-compression: bench-compress crashes due to xz being read-only.
	- lib-lua: Fix linking libdict_lua for non-GNU linkers when Lua support
	  is disabled.
	- lib-mail: There was no limit on how large an email header name could be.
	  Processable header names are now limited to 1000 bytes.
	- lib-oauth2: Dovecot disallowed JWT tokens if their validity time was
	  older than token creation time (nbf < iat).
	- lib-storage: Reduce memory footprint of certain storage operations.
	- lib-storage: When listing mailboxes with storage name escape
	  characters (^ or .) as part of the mailbox name, the listing could
	  show corrupted mailbox names. Due to an issue in handling escaped
	  parent folders, the listing of other mailbox names would become
	  corrupted by prepending parts of the previously listed mailboxes
	  parent folder as prefix to the actual mailbox names. The corruption
	  can occur when using LAYOUT=INDEX and maildir or obox, or when using
	  the listescape plugin.
	- mail-crypt: Fix "-O" argument for "doveadm mailbox cryptokey password"
	  command to be a boolean, and not expect a string.
	- submission-login: Add support for not authenticating to next hop in
	  submission proxying.
	- submission-login: EHLO was not sent again after XCLIENT when doing

src/doveadm/doveadm-replicator.c
/* Copyright (c) 2013-2018 Dovecot authors, see the included COPYING file */

#include "lib.h"
#include "ioloop.h"
#include "str.h"
#include "strescape.h"
#include "istream.h"
#include "write-full.h"
#include "master-service.h"
#include "doveadm.h"
#include "doveadm-print.h"

#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>

struct replicator_context {
	const char *socket_path;
	const char *priority;
	const char *user_mask, *username;
	struct istream *input;
	bool full_sync;
};

extern struct doveadm_cmd_ver2 doveadm_cmd_replicator[];

static void replicator_cmd_help(const struct doveadm_cmd_ver2 *cmd) ATTR_NORETURN;

static void
replicator_send(struct replicator_context *ctx, const char *data)
{
	if (write_full(i_stream_get_fd(ctx->input), data, strlen(data)) < 0)
		i_fatal("write(%s) failed: %m", ctx->socket_path);
}

static void replicator_connect(struct replicator_context *ctx)
{
#define REPLICATOR_HANDSHAKE "VERSION\treplicator-doveadm-client\t1\t0\n"
	const char *line;
	int fd;

	fd = doveadm_connect(ctx->socket_path);
	net_set_nonblock(fd, FALSE);

	ctx->input = i_stream_create_fd_autoclose(&fd, SIZE_MAX);
	replicator_send(ctx, REPLICATOR_HANDSHAKE);

	alarm(5);
	line = i_stream_read_next_line(ctx->input);
	alarm(0);
	if (line == NULL) {
		if (ctx->input->stream_errno != 0) {
			i_fatal("read(%s) failed: %s", ctx->socket_path,
				i_stream_get_error(ctx->input));
		} else if (ctx->input->eof)
			i_fatal("%s disconnected", ctx->socket_path);
		else
			i_fatal("read(%s) timed out", ctx->socket_path);
	}
	if (!version_string_verify(line, "replicator-doveadm-server", 1)) {
		i_fatal_status(EX_PROTOCOL,
			       "%s not a compatible replicator-doveadm socket",
			       ctx->socket_path);
	}
}

static void replicator_disconnect(struct replicator_context *ctx)
{
	if (ctx->input->stream_errno != 0) {
		i_fatal("read(%s) failed: %s", ctx->socket_path,
			i_stream_get_error(ctx->input));
	}
	i_stream_destroy(&ctx->input);
}

static struct replicator_context *
cmd_replicator_init(struct doveadm_cmd_context *cctx)
{
	struct replicator_context *ctx;

	ctx = t_new(struct replicator_context, 1);
	ctx->socket_path = t_strconcat(doveadm_settings->base_dir,
				       "/replicator-doveadm", NULL);

	(void)doveadm_cmd_param_str(cctx, "socket-path", &ctx->socket_path);
	(void)doveadm_cmd_param_bool(cctx, "full-sync", &ctx->full_sync);
	(void)doveadm_cmd_param_str(cctx, "priority", &ctx->priority);
	(void)doveadm_cmd_param_str(cctx, "user-mask", &ctx->user_mask);
	(void)doveadm_cmd_param_str(cctx, "user", &ctx->username);

	replicator_connect(ctx);
	return ctx;
}

static const char *time_ago(time_t t)
{
	int diff = ioloop_time - t;

	if (t == 0)
		return "-";
	return t_strdup_printf("%02d:%02d:%02d", diff/3600,
			       (diff/60)%60, diff%60);
}

static void cmd_replicator_status_overview(struct replicator_context *ctx)
{
	char *line, *value;

	doveadm_print_init(DOVEADM_PRINT_TYPE_TABLE);
	doveadm_print_header("field", "field",
			     DOVEADM_PRINT_HEADER_FLAG_HIDE_TITLE);
	doveadm_print_header("value", "value",
			     DOVEADM_PRINT_HEADER_FLAG_HIDE_TITLE);

	replicator_send(ctx, "STATUS\n");
	while ((line = i_stream_read_next_line(ctx->input)) != NULL) {
		if (*line == '\0')
			break;
		value = strchr(line, '\t');
		if (value != NULL)
			*value++ = '\0';
		else
			value = "";
		doveadm_print(line);
		doveadm_print(value);
	}
	replicator_disconnect(ctx);
}

static void cmd_replicator_status(struct doveadm_cmd_context *cctx)
{
	struct replicator_context *ctx;
	const char *line, *const *args;
	time_t last_fast, last_full, last_success;

	ctx = cmd_replicator_init(cctx);
	if (ctx->user_mask == NULL) {
	        cmd_replicator_status_overview(ctx);
		return;
	}

	doveadm_print_init(DOVEADM_PRINT_TYPE_TABLE);
	doveadm_print_header("username", "username",
			     DOVEADM_PRINT_HEADER_FLAG_EXPAND);
	doveadm_print_header_simple("priority");
	doveadm_print_header_simple("fast sync");
	doveadm_print_header_simple("full sync");
	doveadm_print_header_simple("success sync");
	doveadm_print_header_simple("failed");

	replicator_send(ctx, t_strdup_printf("STATUS\t%s\n",
					     str_tabescape(ctx->user_mask)));
	while ((line = i_stream_read_next_line(ctx->input)) != NULL) {
		if (*line == '\0')
			break;
		T_BEGIN {
			args = t_strsplit_tabescaped(line);
			if (str_array_length(args) >= 5 &&
			    str_to_time(args[2], &last_fast) == 0 &&
			    str_to_time(args[3], &last_full) == 0 &&
			    str_to_time(args[5], &last_success) == 0) {
				doveadm_print(args[0]);
				doveadm_print(args[1]);
				doveadm_print(time_ago(last_fast));
				doveadm_print(time_ago(last_full));
				doveadm_print(time_ago(last_success));
				doveadm_print(args[4][0] == '0' ? "-" : "y");
			}
		} T_END;
	}
	if (line == NULL) {
		i_error("Replicator disconnected unexpectedly");
		doveadm_exit_code = EX_TEMPFAIL;
	}
	replicator_disconnect(ctx);
}

static void cmd_replicator_dsync_status(struct doveadm_cmd_context *cctx)
{
	struct replicator_context *ctx;
	const char *line;
	unsigned int i;

	ctx = cmd_replicator_init(cctx);

	doveadm_print_init(DOVEADM_PRINT_TYPE_TABLE);
	doveadm_print_header("username", "username",
			     DOVEADM_PRINT_HEADER_FLAG_EXPAND);
	doveadm_print_header_simple("type");
	doveadm_print_header_simple("status");

	replicator_send(ctx, "STATUS-DSYNC\n");
	while ((line = i_stream_read_next_line(ctx->input)) != NULL) {
		if (*line == '\0')
			break;
		T_BEGIN {
			const char *const *args = t_strsplit_tabescaped(line);

			for (i = 0; i < 3; i++) {
				if (args[i] == NULL)
					break;
				doveadm_print(args[i]);
			}
			for (; i < 3; i++)
				doveadm_print("");
		} T_END;
	}
	replicator_disconnect(ctx);
}

static void cmd_replicator_replicate(struct doveadm_cmd_context *cctx)
{
	struct replicator_context *ctx;
	string_t *str;
	const char *line;

	ctx = cmd_replicator_init(cctx);
	if (ctx->user_mask == NULL)
		replicator_cmd_help(cctx->cmd);

	str = t_str_new(128);
	str_append(str, "REPLICATE\t");
	if (ctx->priority == NULL)
		str_append_tabescaped(str, "low");
	else
		str_append_tabescaped(str, ctx->priority);
	str_append_c(str, '\t');
	if (ctx->full_sync)
		str_append_c(str, 'f');
	str_append_c(str, '\t');
	str_append_tabescaped(str, ctx->user_mask);
	str_append_c(str, '\n');
	replicator_send(ctx, str_c(str));

	doveadm_print_init(DOVEADM_PRINT_TYPE_FLOW);
	doveadm_print_header("result", "result",
			     DOVEADM_PRINT_HEADER_FLAG_HIDE_TITLE);

	line = i_stream_read_next_line(ctx->input);
	if (line == NULL) {
		i_error("Replicator disconnected unexpectedly");
		doveadm_exit_code = EX_TEMPFAIL;
	} else if (line[0] != '+') {
		i_error("Replicator failed: %s", line+1);
		doveadm_exit_code = EX_USAGE;
	} else {
		doveadm_print(t_strdup_printf("%s users updated", line+1));
	}
	replicator_disconnect(ctx);
}

static void cmd_replicator_add(struct doveadm_cmd_context *cctx)
{
	struct replicator_context *ctx;
	string_t *str;
	const char *line;

	ctx = cmd_replicator_init(cctx);
	if (ctx->user_mask == NULL)
		replicator_cmd_help(cctx->cmd);

	str = t_str_new(128);
	str_append(str, "ADD\t");
	str_append_tabescaped(str, ctx->user_mask);
	str_append_c(str, '\n');
	replicator_send(ctx, str_c(str));

	line = i_stream_read_next_line(ctx->input);
	if (line == NULL) {
		i_error("Replicator disconnected unexpectedly");
		doveadm_exit_code = EX_TEMPFAIL;
	} else if (line[0] != '+') {
		i_error("Replicator failed: %s", line+1);
		doveadm_exit_code = EX_USAGE;
	}
	replicator_disconnect(ctx);
}

static void cmd_replicator_remove(struct doveadm_cmd_context *cctx)
{
	struct replicator_context *ctx;
	string_t *str;
	const char *line;

	ctx = cmd_replicator_init(cctx);
	if (ctx->username == NULL)
		replicator_cmd_help(cctx->cmd);

	str = t_str_new(128);
	str_append(str, "REMOVE\t");
	str_append_tabescaped(str, ctx->username);
	str_append_c(str, '\n');
	replicator_send(ctx, str_c(str));

	line = i_stream_read_next_line(ctx->input);
	if (line == NULL) {
		i_error("Replicator disconnected unexpectedly");
		doveadm_exit_code = EX_TEMPFAIL;
	} else if (line[0] != '+') {
		i_error("Replicator failed: %s", line+1);
		doveadm_exit_code = EX_USAGE;
	}
	replicator_disconnect(ctx);
}

struct doveadm_cmd_ver2 doveadm_cmd_replicator[] = {
{
	.name = "replicator status",
	.cmd = cmd_replicator_status,
	.usage = "[-a <replicator socket path>] [<user mask>]",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('a', "socket-path", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('\0', "user-mask", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
},
{
	.name = "replicator dsync-status",
	.cmd = cmd_replicator_dsync_status,
	.usage = "[-a <replicator socket path>]",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('a', "socket-path", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAMS_END
},
{
	.name = "replicator replicate",
	.cmd = cmd_replicator_replicate,
	.usage = "[-a <replicator socket path>] [-f] [-p <priority>] <user mask>",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('a', "socket-path", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('f', "full-sync", CMD_PARAM_BOOL, 0)
DOVEADM_CMD_PARAM('p', "priority", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('\0', "user-mask", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
},
{
	.name = "replicator add",
	.cmd = cmd_replicator_add,
	.usage = "[-a <replicator socket path>] <user mask>",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('a', "socket-path", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('\0', "user-mask", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
},
{
	.name = "replicator remove",
	.cmd = cmd_replicator_remove,
	.usage = "[-a <replicator socket path>] <username>",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('a', "socket-path", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('\0', "user", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
},
};

static void replicator_cmd_help(const struct doveadm_cmd_ver2 *cmd)
{
	unsigned int i;

	for (i = 0; i < N_ELEMENTS(doveadm_cmd_replicator); i++) {
		if (doveadm_cmd_replicator[i].cmd == cmd->cmd)
			help_ver2(&doveadm_cmd_replicator[i]);
	}
	i_unreached();
}

void doveadm_register_replicator_commands(void)
{
	unsigned int i;

	for (i = 0; i < N_ELEMENTS(doveadm_cmd_replicator); i++)
		doveadm_cmd_register_ver2(&doveadm_cmd_replicator[i]);
}
